### PR TITLE
Adding `rebase-snap` to `snapshot-editor`

### PR DESCRIPTION
**Pull request by ShadowCurse**
## Changes
Added `rebase-snap` function to the `snapshot-editor` tool.

## License Acceptance

By submitting this pull request, I confirm that my contribution is made under
the terms of the Apache 2.0 license. For more information on following
Developer Certificate of Origin and signing off your commits, please check
[`CONTRIBUTING.md`][3].

## PR Checklist

- [ ] If a specific issue led to this PR, this PR closes the issue.
- [x] The description of changes is clear and encompassing.
- [x] Any required documentation changes (code and docs) are included in this PR.
- [ ] API changes follow the [Runbook for Firecracker API changes][2].
- [x] User-facing changes are mentioned in `CHANGELOG.md`.
- [ ] All added/changed functionality is tested.
- [ ] New `TODO`s link to an issue.
- [x] Commits meet [contribution quality standards](https://github.com/firecracker-microvm/firecracker/blob/main/CONTRIBUTING.md#contribution-quality-standards).

---

- [ ] This functionality cannot be added in [`rust-vmm`][1].

[1]: https://github.com/rust-vmm
[2]: https://github.com/firecracker-microvm/firecracker/blob/main/docs/api-change-runbook.md
[3]: https://github.com/firecracker-microvm/firecracker/blob/main/CONTRIBUTING.md

### CODE REVIEW

File: docs/snapshotting/snapshot-editor.md [hunk 1]
```diff
@@ -10,11 +10,39 @@ Firecracker snapshot consists of 2 files:
   devices states etc.
 - `memory` file: file with guest memory.
 
-`snapshot-editor` (currently) allows to modify only `vmstate` files only
-on aarch64 platform.
-
 ## Usage
 
+### `edit-memory` command
```
[the remark this answers is not in the document]
Comment (ShadowCurse) [reply docs/snapshotting/snapshot-editor.md:15]: I am not sure enumerating initial features is necessary.
It would look like this:
- [#4045](https://github.com/firecracker-microvm/firecracker/pull/4045)
  Added `snapshot-editor` tool for modifications of snapshot files.
  Initial commands/subcommands are:
    - `edit-memory` command:
      - `rebase` subcommand
    - `edit-vmstate` command:
      - `remove-regs` subcommand (aarch64 only)
    - `info-vmstate` command:
      - `version` subcommand
      - `vcpu-states` subcommand (aarch64 only)

File: src/snapshot-editor/src/edit_memory.rs [hunk 2]
```diff
@@ -0,0 +1,234 @@
...
+        let base = tempfile::TempFile::new().unwrap();
+        let diff = tempfile::TempFile::new().unwrap();
+
+        let mut base_file = base.as_file();
+        let mut diff_file = diff.as_file();
+
+        let base_path = base.as_path().to_path_buf();
+        let diff_path = diff.as_path().to_path_buf();
+
+        // 1. Empty files
```
[the remark this answers is not in the document]
Comment (ShadowCurse) [reply src/snapshot-editor/src/edit_memory.rs]: split corner cases test into several

File: docs/snapshotting/snapshot-editor.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (kalyazin) [reply docs/snapshotting/snapshot-editor.md:15]: yeah, I mostly envisioned it as a plain text like `Added snapshot-editor tool for modifications of snapshot files allowing rebasing memory snapshot files, printing and removing aarch64 registers from the vmstate`. Listing detailed subcommands in the changelog is certainly an overkill.

**Review by kalyazin — APPROVED** [review]

File: src/snapshot-editor/src/edit_memory.rs [hunk 3]
```diff
@@ -0,0 +1,257 @@
...
+            // diff:      [ ] ___ [0] [ ]
+            // expected:  [d] [b] [d] [d]
+            let diff_block = rand::rand_bytes(block_size);
+            diff_file.write_all(&diff_block).unwrap();
+            expected_result.extend(diff_block);
+            // Rebase and check the result
+            rebase(base_path.clone(), diff_path.clone()).unwrap();
+            check_file_content(base_file, &expected_result);
+
+            // 5. The base file is bigger
```
Comment (sudanl0) [inline src/snapshot-editor/src/edit_memory.rs:242]: thanks for making the use cases in the test more clear with the visual comments!

**Review by sudanl0 — APPROVED** [review]

File: src/snapshot-editor/src/edit_memory.rs [hunk 4]
```diff
@@ -0,0 +1,257 @@
...
+use std::path::PathBuf;
+
+use clap::Subcommand;
+use fc_utils::seek_hole::SeekHole;
+
+#[derive(Debug, thiserror::Error)]
+pub enum EditMemoryError {
+    #[error("Could not open memory file: {0}")]
+    OpenMemoryFile(std::io::Error),
+    #[error("Could not open memory file: {0}")]
```
[the remark this answers is not in the document]
Comment (ShadowCurse) [reply src/snapshot-editor/src/edit_memory.rs]: fixed

File: src/snapshot-editor/src/edit_memory.rs [hunk 5]
```diff
@@ -0,0 +1,257 @@
...
+}
+
+#[derive(Debug, Subcommand)]
+pub enum EditMemorySubCommand {
+    /// Remove registers from vcpu states.
+    Rebase {
+        /// Path to the memory file.
+        #[arg(short, long)]
+        memory_path: PathBuf,
+        /// Path to the memory file.
```
[the remark this answers is not in the document]
Comment (ShadowCurse) [reply src/snapshot-editor/src/edit_memory.rs]: fixed

File: docs/snapshotting/snapshot-editor.md [hunk 6]
```diff
@@ -10,11 +10,39 @@ Firecracker snapshot consists of 2 files:
...
+
+#### `rebase` subcommand
+
+This command is used to merge a `diff` snapshot memory file on
+top of a base memory file.
+
+**None**
```
[the remark this answers is not in the document]
Comment (ShadowCurse) [reply docs/snapshotting/snapshot-editor.md]: It should have been `Note`. Fixed

**Review by sudanl0 — APPROVED** [review]

**Review by kalyazin — APPROVED** [review]
